jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/BuildImageStep.java

```diff
@@ -94,22 +96,47 @@
     try (Timer ignored = new Timer(buildConfiguration.getBuildLogger(), DESCRIPTION)) {
       // Constructs the image.
       Image.Builder<CachedLayer> imageBuilder = Image.builder();
-      for (PullAndCacheBaseImageLayerStep pullAndCacheBaseImageLayerStep :
-          NonBlockingSteps.get(pullAndCacheBaseImageLayersStep)) {
+      Image<Layer> baseImage = NonBlockingSteps.get(pullBaseImageStep).getBaseImage();
+      ContainerConfiguration containerConfiguration =
+          buildConfiguration.getContainerConfiguration();
+
+      // Base image layers
+      List<PullAndCacheBaseImageLayerStep> baseImageLayers =
+          NonBlockingSteps.get(pullAndCacheBaseImageLayersStep);
+      for (PullAndCacheBaseImageLayerStep pullAndCacheBaseImageLayerStep : baseImageLayers) {
         imageBuilder.addLayer(NonBlockingSteps.get(pullAndCacheBaseImageLayerStep));
       }
-      for (BuildAndCacheApplicationLayerStep buildAndCacheApplicationLayerStep :
-          buildAndCacheApplicationLayerSteps) {
-        imageBuilder.addLayer(NonBlockingSteps.get(buildAndCacheApplicationLayerStep));
-      }
 
-      // Parameters that we passthrough from the base image
-      Image<Layer> baseImage = NonBlockingSteps.get(pullBaseImageStep).getBaseImage();
+      // Passthrough config and count non-empty history entries
+      int nonEmptyLayerCount = 0;
+      for (HistoryItem historyObject : baseImage.getHistory()) {
+        imageBuilder.addHistory(historyObject);
+        if (!historyObject.isEmptyLayer()) {
+          nonEmptyLayerCount++;
+        }
+      }
       imageBuilder.addEnvironment(baseImage.getEnvironment());
       imageBuilder.addLabels(baseImage.getLabels());
 
-      ContainerConfiguration containerConfiguration =
-          buildConfiguration.getContainerConfiguration();
+      // Add history elements for non-empty layers that don't have one yet
+      int sizeDifference =
+          baseImageLayers.size() + buildAndCacheApplicationLayerSteps.size() - nonEmptyLayerCount;
+      for (int count = 0; count < sizeDifference; count++) {
```

Comment:
My only concern here is that we're claiming (and setting ourselves up for blame for) layers that were not generated by us.  Perhaps we could only add missing-history items here, and then add jib layers in the next `for()` where you're adding the Jib-generated layers (lines 136–139)?

Reply:
Yeah, I originally had it this way and changed it so we only needed to use `new HistoryItem()` once, but on second thought I agree it's clearer the original way (easier to tell which history entries correspond to which layers).